### ISSUE TITLE
Performance of a direct vs sequential union_all

### ISSUE THREAD

**Issue by martinfleis** [pygeos/pygeos#292]
Hi,

I have to dissolve a large number (millions) of dense polygons and trying to figure a way how to parallelize it. During that, I have encountered a strange behaviour of `union_all` performance (which I guess is inherited from GEOS).

It seems that using `union_all` directly on the whole array is the slowest possible option. If I do the union of parts and that merge parts together (which results in the same polygon) I can get there up to 15x faster (and maybe even more). Consider the example below, where I check different sizes of chunks for the initial sequential union.

With large arrays, it may be even worth doing the chunk-based union twice in a cascaded way. What puzzles me is why is that and why something like this is not implemented directly in GEOS since it can provide a significant performance boost. Do you have any idea? And any idea how to come with a heuristic for this (which can be then parallelised)? Even splitting the array into two parts can give 2x speedup. I haven't found anything describing this behaviour.

If we figure out a way how to manage this we can even reimplement it as a custom performant version of `union_all`.

The time to do `pygeos.union_all(geoms)` directly is 131s.

```py
from time import time
import geopandas as gpd
import pygeos

sample = gpd.read_parquet("https://www.dropbox.com/s/oxy2h2pb0m3no5d/sample.pq?dl=1")

geoms = sample.geometry.values.data

times = pd.DataFrame(columns=['time'])

size = geoms.shape[0]

for split in [2, 5, 10, 25, 50, 100, 150, 200, 350, 500, 1000, 1500]:
    s = time()
    unoined = pygeos.union_all(
        [pygeos.union_all(geoms[edge : edge + split]) for edge in range(0, size, split)]
    )
    times.loc[split] = time() - s

times.plot(xlabel="split size", ylabel='time')
```
![Screenshot 2021-02-02 at 17 22 30](https://user-images.githubusercontent.com/36797143/106637952-3cc72900-657b-11eb-9e10-bcdbc8add892.png)

**Comment by brendan-ward** [comment]
This is interesting!  I've been trying to optimize dissolve too, with a different approach.  Thank you for pulling together a clear case to demonstrate variation in performance based on chunking this up.

Looks like all geometries are part of one adjacent group, so using an index to dissolve adjacent geometries won't help here:
![image](https://user-images.githubusercontent.com/3375604/106639597-fac9c300-6539-11eb-8f54-8d335ebeeee8.png)

Unfortunately, it looks like `coverage_union_all` fails for these data, even though geometries share edges (must not be topologically perfect); I've found that where topology is clean, that can give significant speedups.

**Comment by jorisvandenbossche** [comment]
It also reminds me of https://github.com/pygeos/pygeos/issues/244, although it is "the other way around", about subdividing a large geometry into smaller geometries to get faster operations. 
(depending on the implementation of UnaryUnion in GEOS, it could actually be related, as when unioning one huge array, you are dealing with one huge end result for all geometries, while each of the intermediate resulting geometries are smaller and thus faster to work with)

**Comment by brendan-ward** [comment]
I thought GEOS was using cascaded union behind unary union, which I thought used a spatial index to help iteratively aggregate geometries.  But I may have that wrong.  If it doesn't use cascaded union, we may want to expose that in pygeos for situations like this.

@martinfleis what version of pygeos and GEOS are you using here?

I'm not seeing quite the same dramatic difference with your test data, on `master` of both pygeos and GEOS (>3.9).
![sample_time](https://user-images.githubusercontent.com/3375604/106643161-28b10680-653e-11eb-825d-bd49d5a917ed.png)

**Comment by brendan-ward** [comment]
Not sure if it helps much in this case, but I noticed that you have a mix of single and mulipolygons.  Exploding them first via `pygeos.get_parts` may help improve things slightly, esp. if there is a spatial index used to merge polygons as part of union.

![sample_time](https://user-images.githubusercontent.com/3375604/106644470-ff917580-653f-11eb-9ac7-a26b40fe80d1.png)

**Comment by martinfleis** [comment]
The timing above is with pygeos 0.8 I had in the container and GEOS 3.8.1. PyGEOS master and GEOS 3.8.1 is the same. I'll try with GEOS 3.9, maybe something changed under the hood there.

@brendan-ward your times are much much smaller than mine, are you using different data or a sample of that parquet?

> If it doesn't use cascaded union, we may want to expose that in pygeos for situations like this.

While I was trying to find more info about this I learned that cascaded union is actually deprecated in GEOS. Shapely uses unary_union when cascaded_union is called (to be deprecated in 1.8 I think).

Using spatial index would help only if you have more disconnected components right? Which is often the case for dissolve but not necessarily here.

**Comment by brendan-ward** [comment]
@martinfleis I'm using the sample file you posted above with 2,285 records on MacOS with Python 3.8.6.

I think there are 2 ways in which a spatial index could be used for union:
1) the approach that I understand to be used by cascaded union ([more here](http://blog.cleverelephant.ca/2009/01/must-faster-unions-in-postgis-14.html)) puts everything into a spatial index, then aggregates starting from the leaves and works upward toward the root of the tree.
2) using a spatial index _before_ calling `union_all` (i.e., outside GEOS) to define groups of adjacent geometries, when the underlying geometries have some groupings and some isolated geometries.  Where geometries are isolated, those can be grouped back together using `pygeos.multipolygons` (if they are single part polys) which is faster than calling union on those same geoms.  The adjacent groups can then each individually dissolved and added to the result.  I'm still tuning this approach; it seems to work reasonably well for some of my real-world data, but not so much for my test data that I'm using to try and prove the algorithm.  The gains of doing this are proportional to the amount of isolated geometries, but there is some overhead to building the index and adjacency matrix ourselves.

**Comment by jorisvandenbossche** [comment]
> I'll try with GEOS 3.9, maybe something changed under the hood there.

The overlay operations saw a big overhaul with OverlayNG, so it seems this has been largely solved in those improvements.

With GEOS 3.8.1

![Figure_1_38](https://user-images.githubusercontent.com/1020496/106646999-a997ee80-658e-11eb-99f8-4aead8d5ca19.png)

With GEOS 3.9.0

![test39](https://user-images.githubusercontent.com/1020496/106646970-9ab13c00-658e-11eb-9c14-fc314d6af1b8.png)

**Comment by martinfleis** [comment]
Wow! That is insane. Just got the same result here. That means ~40x speedup between 3.8.1 and 3.9. I guess my today's effort, although interesting is no longer relevant :D.

**Comment by darkblue-b** [comment]
:100:

**Comment by dr-jts** [comment]
> What puzzles me is why is that and why something like this is not implemented directly in GEOS since it can provide a significant performance boost. 

GEOS does implement a cascaded unary union approach. 

But it doesn't do it in parallel, so that's something that can be done in `pygeos`.  Or is that what `union_all` is doing already?

**Comment by jorisvandenbossche** [comment]
> But it doesn't do it in parallel, so that's something that can be done in `pygeos`. Or is that what `union_all` is doing already?

PyGEOS itself doesn't do that, currently. But it does release the GIL, which means that another python applications can use call pygeos in parallel from in multiple threads (and this for example what dask-geopandas experiments with)

**Comment by martinfleis** [comment]
I am going to close this issue as resolved by GEOS 3.9. There is still some difference but it is not stable (I get a slightly different outcome every run) and it oscillates in ± 20% margin. Thank you, everyone, for a fascinating discussion!
